EIPS/eip-2381.md

```diff
@@ -0,0 +1,143 @@
+---
+eip: 2381
+title: ENS support for EIP-721 TokenID
+author: Oisin Kyne (@OisinKyne)
+discussions-to: https://github.com/ethereum/EIPs/issues/2381
+status: Draft
+type: Standards Track
+category: ERC
+created: 2019-11-15
+requires: 137, 165, 721
+---
+
+## Simple Summary
+
+This EIP makes [EIP-721](https://eips.ethereum.org/EIPS/eip-721) Non-Fungible Tokens addressable by the Ethereum Name Service ([EIP-137](https://eips.ethereum.org/EIPS/eip-137)) by adding a new resolver profile, `uint256 tokenID()`. This allows the Ethereum Name Service to address a single non-fungible token within an EIP-721 token contract of many tokens.
+
+## Abstract
+
+It makes sense to resolve an ENS name to just a contract address for fungible tokens such as [EIP-20](https://eips.ethereum.org/EIPS/eip-20), as each token in the contract is indistinguishable from another. However, for non-fungible tokens; pointing to the contract address alone is not enough, as tokens within the contract are meant to be unique and distinguishable. Non-fungibles might have different valuations, might grant the holder different rights or rewards etc.
+
+A single deployed EIP-721 contract can contain multiple non-fungible tokens. Tokens are addressed by an unsigned 256bit integer, `tokenID`. This EIP describes an extra getter and setter, `tokenID()` and `setTokenID()`, that are included in an ENS resolver contract, to allow for ENS names to resolve to a specific non fungible token within a given EIP-721 contract.
+
+This allows ENS domains like [`bugcat.cryptokitties.eth`](https://www.cryptokitties.co/kitty/101) and [`dragon.cryptokitties.eth`](https://www.cryptokitties.co/kitty/896775) to resolve to both the CryptoKitties contract address and the tokenID of the non-fungible kitty within it.
+
+## Motivation
+
+Non-fungible Tokens are meant to be globally unique, it makes sense to be able to name them. Giving NFTs names makes them more real to the non Ethereum enthusiast that doesn't understand what a contract address or tokenID is. A domain name in the existing web 2.0 world is already widely understood to be a finite resource, and as such, would help convey the scarcity that is a non-Fungible token to new users.
+
+This change might also to an extent, decentralise access to non-fungibles on Ethereum. Currently, the mapping for a human readable name to an NFT is kept off chain, typically on the platform of the NFT issuer. However, if the community moved towards naming NFTs on chain, using ENS, this would allow any client that supports ENS resolution to resolve an ENS name to an NFT, without querying the issuer's off-chain metadata.
+
+## Specification
+
+The ENS Resolver Profile specified by this EIP adds two functions to a deployed resolver contract:
+
+```
+    function tokenID(bytes32 node) public view returns(uint256);
+    function setTokenID(bytes32 node, uint256 token);
```

Comment:
The specification should probably specify what value `tokenId` contains.  At the moment, if I use this resolver and call `setTokenID(..., 101)` and then someone resolves via `tokenID(...)`, they will not have enough information to get to `https://www.cryptokitties.co/kitty/101` as the motivation suggests.  This implies to me that there is something left unspecified here.

Reply:
Added an explainer to line 21 that the URL is to be retrieved from from EIP721's `tokenURI` functionality, and this EIP is just to give people the tokenID they need to request the right url. https://github.com/ethereum/EIPs/blame/master/EIPS/eip-721.md#L193 hope that makes it clearer.